File: grbASP/python/PsfClusters.py
"""
@brief Use PSF to determine cluster locations and to construct spatial
part of log-likelihood.

@author J. Chiang <jchiang@sla.stanford.edu>
"""
#
# $Header$
#

import numpy as num
import pyIrfLoader
from ScData import ScData
SkyDir = pyIrfLoader.SkyDir

_scData = ScData()

pyIrfLoader.Loader_go()
_irfsFactory = pyIrfLoader.IrfsFactory_instance()

irfs = ('DC2::FrontA', 'DC2::BackA', 'DC2::FrontB', 'DC2::BackB')
#irfs = ('DC1A::Front', 'DC1A::Back', 'DC1A::Front', 'DC1A::Back')
_psfs = [_irfsFactory.create(irf).psf() for irf in irfs]

class PsfClusters(object):
    def __init__(self, events):
        dists = {}
        for evt in events:
            dists[evt] = []
            for other in events:
                dists[evt].append( (evt.sep(other), other) )
        self.dists = dists
    def _clusterSizes(self):
        sizes = {}
        for evt in self.dists:
            sizes[evt] = self._psfLogLike(evt)
        return sizes
    def _psfLogLike(self, evt):
        logLike = 0
        psf_wts = {}
        theta = _scData.inclination(evt.time, evt.dir)
        for sep, other in self.dists[evt]:
            psf = _psfs[evt.evt_class]
            psf_wts[other] = psf.value(sep, other.energy, theta, 0)
            logLike += num.log(psf_wts[other])
        return logLike, psf_wts
    def _largestCluster(self, radius=None):
        sizes = self._clusterSizes()
        maxval = sizes[sizes.keys()[0]]
        for evt in sizes:
            if maxval[0] < sizes[evt][0]:
                largest = evt
                maxval = sizes[evt]
        return maxval
    def logLike(self, radius=None, bg_rate=None):
        times = []
        for evt in self.dists:
            times.append(evt.time)
        times.sort()
        times = num.array(times)
        dts = times[1:] - times[:-1]
        if bg_rate is None:
            bg_rate = len(times)/(times[-1] - times[0])
        xvals = bg_rate*dts
        logPdts = sum(num.log(1. - num.exp(-xvals)))

        max_cluster = self._largestCluster()
        logPdists = -max_cluster[0]
        
        return logPdts, logPdists
    def localize(self, cluster=None):
        if cluster is None:
            logLike, psf_wts = self._largestCluster()
        ra_avg, dec_avg, norm = 0, 0, 0
        ras, decs = [], []
        for evt in psf_wts:
            ras.append(evt.ra)
            decs.append(evt.dec)
            ra_avg += evt.ra*psf_wts[evt]
            dec_avg += evt.dec*psf_wts[evt]
            norm += psf_wts[evt]
        return SkyDir(ra_avg/norm, dec_avg/norm), ras, decs

if __name__ == '__main__':
    import os
    from FitsNTuple import FitsNTuple
    import hippoplotter as plot
    from blind_search import BlindSearch
    
    downlink = os.environ['DOWNLINKFILE']
    events = FitsNTuple(downlink)
    blindSearch = BlindSearch(events, clusterAlg=PsfClusters)
    
    plot.scatter(blindSearch.times, blindSearch.logdts, pointRep='Line')
    plot.scatter(blindSearch.times, blindSearch.logdists, pointRep='Line')
    logLike = blindSearch.logdts + blindSearch.logdists
    plot.scatter(blindSearch.times, logLike, pointRep='Line')

    hist = plot.histogram(events.TIME)
    hist.setBinWidth('x', 5)

    grb_dirs = blindSearch.grbDirs()
    for grb in grb_dirs:
        plot.xyhist(grb[1], grb[2], xname='RA', yname='Dec')
        plot.vline(grb[0].ra())
        plot.hline(grb[0].dec())

File: grbASP/src/PsfClusters.cxx
/**
 * @file PsfClusters.cxx
 * @brief Photon cluster detection algorithm using the PSF
 * @author J. Chiang <jchiang@slac.stanford.edu>
 *
 * $Header$
 */

#include <iostream>
#include <sstream>
#include <stdexcept>
#include <utility>

#include "irfInterface/IrfsFactory.h"
#include "irfInterface/IPsf.h"

#include "irfLoader/Loader.h"

#include "grbASP/PsfClusters.h"
#include "grbASP/ScData.h"

namespace grbASP {

PsfClusters::PsfClusters(const std::vector< std::pair<double, double> > & gtis,
                         const ScData & scData,
                         const std::string & irfs) 
   : ClusterAlg(gtis), m_scData(scData) {
   loadIrfs(irfs);
}

void PsfClusters::process(const std::vector<Event> & events,
                          double & logLike_time, double & logLike_pos,
                          astro::SkyDir & meanDir, double bg_rate) {
   m_events = events;
   logLike_time = logLikeTime(bg_rate);

   std::map<Event, double> wts;
   const Event & largest(findLargestCluster(logLike_pos));
   logLike_pos *= -1;
   computePsfWts(largest, wts);
   double xhat(0);
   double yhat(0);
   double zhat(0);
   double norm(0);
   for (std::map<Event, double>::const_iterator wt(wts.begin());
        wt != wts.end(); ++wt) {
      xhat += wt->first.dir().dir().x()*wt->second;
      yhat += wt->first.dir().dir().y()*wt->second;
      zhat += wt->first.dir().dir().z()*wt->second;
      norm += wt->second;
   }
   xhat /= norm;
   yhat /= norm;
   zhat /= norm;
   meanDir = astro::SkyDir(CLHEP::Hep3Vector(xhat, yhat, zhat));
}

PsfClusters::~PsfClusters() {
   for (std::map<size_t, irfInterface::IPsf *>::iterator item(m_psfs.begin());
        item != m_psfs.end(); ++item) {
      delete item->second;
      item->second = 0;
   }
}

double PsfClusters::eventLogLike(const Event & evt) const {
   std::map<Event, double> wts;
   computePsfWts(evt, wts);
   return eventLogLike(wts);
}

double PsfClusters::eventLogLike(const std::map<Event, double> & wts) const {
   double logLike(0);
   for (std::map<Event, double>::const_iterator wt(wts.begin());
        wt != wts.end(); ++wt) {
      logLike += std::log(wt->second);
   }
   return logLike;
}

const Event & PsfClusters::findLargestCluster(double & maxLogLike) const {
   size_t ilargest(0);
   maxLogLike = eventLogLike(m_events.front());
   for (size_t i(1); i < m_events.size(); i++) {
      double currentLogLike(eventLogLike(m_events.at(i)));
      if (currentLogLike > maxLogLike) {
         ilargest = i;
         maxLogLike = currentLogLike;
      }
   }
   return m_events.at(ilargest);
}

void PsfClusters::computePsfWts(const Event & evt, 
                                std::map<Event, double> & wts) const {
   wts.clear();
   double theta(m_scData.inclination(evt.time(), evt.dir()));
   std::vector<Event>::const_iterator event(m_events.begin());
   for ( ; event != m_events.end(); ++event) {
      double sep(event->sep(evt));
      double value(psf(event->eventClass()).value(sep, event->energy(),
                                                  theta, 0));
      wts.insert(std::make_pair(*event, value));
   }
}

void PsfClusters::loadIrfs(const std::string & irfs) {
   irfLoader::Loader::go(irfs);
   const std::vector<std::string> & irfNames =
      irfLoader::Loader::respIds().find(irfs)->second;
   std::cerr << "Using IRFs: " << std::endl;
   for (size_t i(0); i < irfNames.size(); i++) {
      std::cerr << "  " << irfNames.at(i) << std::endl;
      irfInterface::Irfs * irf = 
         irfInterface::IrfsFactory::instance()->create(irfNames.at(i));
      m_psfs[irf->irfID()] = irf->psf()->clone();
      delete irf;
   }
}

const irfInterface::IPsf & PsfClusters::psf(size_t item) const {
   typedef std::map<size_t, irfInterface::IPsf *> PsfMap_t;
   PsfMap_t::const_iterator it(m_psfs.find(item));
   if (it == m_psfs.end()) {
      std::cout << "Available IRF classes: \n";
      for (PsfMap_t::const_iterator foo(m_psfs.begin()); 
           foo != m_psfs.end(); ++foo) {
         std::cout << foo->first << std::endl;
      }
      std::ostringstream message;
      message << "Error accessing IRF for event class " << item;
      throw std::runtime_error(message.str());
   }
   return *(it->second);
}

} // namespace grbASP
